gnovm/pkg/gnolang/realm.go

```diff
@@ -876,6 +883,145 @@ func (rlm *Realm) clearMarks() {
 	rlm.escaped = nil
 }
 
+// assertObjectIsPublic ensures that the object is public and does not have any private dependencies.
+// it check recursively the types of the object
+// it does not recursively check the values because
+// child objects are validated separately during the save traversal (saveUnsavedObjectRecursively)
+func (rlm *Realm) assertObjectIsPublic(obj Object, store Store, visited map[TypeID]struct{}) {
```

Comment:
can't these check happen in `DidUpdate`?

Reply:
i didn't see that, can you double check?

Reply:
sorry, i had a problem with github and i deleted some older comment but IIRC i was saying that i added these check in DidUpdate first and @thehowl advised it was better to make it happen in FinalizeRealm since DidUpdate is called many times from different places, leading to potentiel useless checks. wdyt ? why these check should happen in DidUpdate ? from my experience, i don't see major pros & cons and i agree finalizeRealm looks like a better place imo

Reply:
an object is firstly passed in `DidUpdate` and then the `FinalizeRealmTransaction`, which consists several different phases handling before it's finally persisted. that said, there's no less call doing the check in `FinalizeRealmTransaction` than in `DidUpdate`, right?  if so, we can always check `earlier` and fail `earlier`, will that be better? So, i guess maybe the check can happen in `DidUpdate` or even before that, like what already have in the `ReadOnly` stuff.

Reply:
yes, moving it to FinalizeRealmTransaction was on my suggestion @ltzmaxwell.

I recall earlier discussions with Jae where it was said that DidUpdate should be kept mean & lean

Reply:
cc: @jaekwon for post review.